src/librustc/ty/mod.rs

```diff
@@ -994,8 +994,12 @@ impl<'a, 'gcx, 'tcx> Predicate<'tcx> {
 }
 
 #[derive(Clone, Copy, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
+pub enum DefaultImplCheck { Yes, No, }
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
 pub struct TraitPredicate<'tcx> {
-    pub trait_ref: TraitRef<'tcx>
+    pub trait_ref: TraitRef<'tcx>,
+    pub default_impl_check: DefaultImplCheck
```

Comment:
What is the function of this field?

Reply:
This field is like a flag used during Trait Resolution.
When setted to `Yes`, the `ImplCandidate` resolved from an `Obligation` that contain the `TraitPredicate` are checked to see if they refer to a `default impl`.
An `Err(Unimplemented)` is returned in that case because a `default impl` doesn't implement a `Trait`.
For example, given this code 
```
trait Foo {
    fn foo_one(&self) -> &'static str;
    fn foo_two(&self) -> &'static str;
}

default impl<T> Foo for T {
    fn foo_one(&self) -> &'static str {
        "generic"
    }
}

struct MyStruct;
```
 - a call like `MyStruct.foo_one()` should pass the trait resolution since the `TraitPredicate(MyStruct as Foo)` is marked as `No`
 - a call like `foo(MyStruct)` with foo like `fn foo<T: Foo>(x: T) -> &'static str` should fail the trait resolution since the `TraitPredicate(MyStruct as Foo)` is marked as `Yes`

Reply:
OK, I don't think we should need this field once we've aligned things correctly.